### ISSUE TITLE
504 Server Error in X-Match query

### ISSUE THREAD

**Issue by cosmicpudding** [astropy/astroquery#1900]
I have been using astroquery's wrapper around X-Match to run crossmatches against big surveys, and strangely it fails for certain surveys which is new behaviour. I see the following error:

`   d = XMatch.query(cat1=open(masterlist),cat2='vizier:%s' % code, max_distance=angular_sep * u.arcsec,colRA1='ra',colDec1='dec')
  File "/Users/moss/anaconda/lib/python3.6/site-packages/astroquery/xmatch/core.py", line 61, in query
    get_query_payload=get_query_payload)
  File "/Users/moss/anaconda/lib/python3.6/site-packages/astroquery/xmatch/core.py", line 95, in query_async
    response.raise_for_status()
  File "/Users/moss/anaconda/lib/python3.6/site-packages/requests/models.py", line 939, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 504 Server Error: Gateway Time-out for url: http://cdsxmatch.u-strasbg.fr/xmatch/api/v1/sync`

I checked the self.TIMEOUT parameter passed to the requests call, which was 300, but I tried setting it to much higher and that made no difference to the error, it still timed out and refused to continue. This was specifically for querying against the ALLWISE catalogue, whereas for other catalogues (e.g. SUMSS, XMMDR8, etc) it finishes within about a minute. It's possible this is not an astroquery issue but traces back to X-Match, but I found when I did a crossmatch on the X-Match website (against 2MASS, an equally big catalogue) it took a long time but actually did seem to work okay. Wondering if there is any kind of fix possible on the astroquery side? Thanks in advance!

**Comment by cosmicpudding** [comment]
p.s. it looks like it also failed for SDSS... so maybe an issue with how it connects to the large catalogues? I did try to look through existing issues but couldn't find anything about this.

**Comment by keflavich** [comment]
@tboch @fxpineau can either of you help with this?

**Comment by fxpineau** [comment]
Thank you @keflavich. 
@cosmicpudding what is the size (number of rows) and the sky distribution (allsky, small area) of your input table?
(I have not changed the xmatch server configuration, so I wonder if this is a new behavior or not)

**Comment by cosmicpudding** [comment]
Thanks @fxpineau and @keflavich for looking at this! The input catalogues for these crossmatches have on order 700,000 rows, and it's pretty much all (southern) sky. I did check the limits and it suggested that if there were more than 2 million matches it would truncate but didn't seem to imply a limit otherwise, and the file size is only 20 MB or so (I think 100 MB is the limit?).

**Comment by cosmicpudding** [comment]
I guess it's new behaviour since the last time I was using this part of the code, which was in April 2020, though last time I did it the number of rows was about half of what it is now, so it's possible that the current table size could be hitting a limit somewhere?

**Comment by fxpineau** [comment]
Thank you @cosmicpudding.
See the edit of my comment below: 700,000 rows and half the sky may be too much for the synchronous service.
Although 100 MB is the upload limit, it accounts for possible columns others than the positions. 
The 2 million is an output limit (in case of large radius and/or not very large tables and/or limited area with a large catalogue).
I would recommend to split the input in chunks of 10,000-50,000 rows (before making the chunks, sorting the rows by their HEALPix index at order e.g. 12 will speed up the xmatch).

**Comment by cosmicpudding** [comment]
Thanks @fxpineau - I can definitely split things into smaller chunks before uploading and will post here once I've done that to confirm how it goes. I haven't worked with HEALPix much before - is there an easy way (e.g. using astropy) to sort rows by HEALPix order? Not entirely sure where to start with that so a pointer would be really helpful!

**Comment by cosmicpudding** [comment]
p.s. I found this: https://astropy-healpix.readthedocs.io/_/downloads/en/v0.4/pdf/ (so probably should be able to figure the HEALPix index thing out) :)

**Comment by cosmicpudding** [comment]
As a random side thought, I've now written code to take the input catalogue, convert it to SkyCoord then HEALPix, add that column to the initial table and then write out various CSVs for each index. Though I'm keeping them, they could easily be temporary CSVs that are generated just for the purpose of uploading to XMatch and then throwing away. Is this something that could be embedded into astroquery.xmatch, so that the user doesn't have to think about dividing it in this way and so you could throw large catalogues at the function without it falling over? I doubt the way I've implemented it is the best way but I can share what I did in case it is helpful. It's a feature I'd like to see! And it's not too complicated to implement (though I am using multiprocessing.Pool to make the calculations go a bit faster).

**Comment by cosmicpudding** [comment]
Two small updates.

1) To give an idea of HEALPix conversion runtime (presumably it could be optimised way better):
`Reading catalogues...
706296 633224
Constructing coordinate list..
Calculating HEALPix using pool...
Adding HEALPix info to table...
Total time: 14.56 min`

2) Breaking the catalogue into 50k chunks worked, and ALLWISE is now returning a response. Thanks @fxpineau!

**Comment by keflavich** [comment]
@cosmicpudding Yes, that sort of functionality would be very welcome!  Can you submit it as a PR?

@fxpineau you mentioned that the original table was too big for the synchronous service.  Is there an asynchronous service we do not yet have an interface to?

**Comment by cosmicpudding** [comment]
@keflavich I can try! I'm not exactly sure where this code should sit in terms of architecture but I'll have a look at what is there already and try to have a go at inserting what I did.

**Comment by fxpineau** [comment]
@keflavich
>  Is there an asynchronous service we do not yet have an interface to?

The jobs submitted on the CDS XMatch web page are executed asynchronously, but so far there is no async HTTP API.

**Comment by HNLala** [comment]
@fxpineau 

> I would recommend to split the input in chunks of 10,000-50,000 rows (before making the chunks, sorting the rows by their HEALPix index at order e.g. 12 should speed up the xmatch).

Thanks for this tip.
Just to be sure, one should use the nested ordering scheme for creating the HEALPix index before sorting the rows?

**Comment by fxpineau** [comment]
> Thanks for this tip.

You are welcome.

> Just to be sure, one should use the nested ordering scheme for creating the HEALPix index before sorting the rows?

Yes: the **nested** scheme has nice indexation properties (since it relies on one z-ordre curve per base cell).
